### PR TITLE
Adds names to LanguageAttribute and makes their codes unique

### PR DESCRIPTION
**Pull request by JoshKeegan**
Pretty much as the title says.
I wanted to have the names of languages in the attribute to let them be displayed to the end-user in a friendlier way than using their tag or ToString() on the enum value.

Whilst doing that, I noticed there were several duplicate entries (different entries for the same language code where a language has multiple common names) which I have removed as it raises the question of which one is correct to return when fetching a Language enum entry frmo its code.

There are also other minor changes e.g. renaming Poi => HighlandPopoluca to be consistent with the naming convention.

### CODE REVIEW

**Comment by JoshKeegan** [comment]
Oh and I forgot to mention the change to LanguageAttribute where Language and Languages have been renamed to Code and Codes accordingly so as to be clear what it represents now LanguageAttribute also stores the name of the language.

**Comment by linvi** [comment]
I am currently reviewing your changes but I wanted to understand why you changed things like :

* Removed `Ivrit`
* Changed `Inupiaq` to `Inupiat`

I will comment on the pull request so that it is easier for you to see what I mean.

File: Tweetinvi.Core/Public/Models/Enum/Language.cs [hunk 1]
```diff
@@ -3,233 +3,198 @@
...
-        [Language("ha")] Hausa,
-        [Language("he")] Hebrew,
-        [Language("hz")] Herero,
-        [Language("hi")] Hindi,
-        [Language("ho")] HiriMotu,
-        [Language("hu")] Hungarian,
-        [Language("ia")] Interlingua,
-        [Language("id")] Indonesian,
-        [Language("ie")] Interlingue,
-        [Language("iw")] Ivrit,
```
Comment (linvi) [inline Tweetinvi.Core/Public/Models/Enum/Language.cs]: Why has this one been removed?

File: Tweetinvi.Core/Public/Models/Enum/Language.cs [hunk 2]
```diff
@@ -3,233 +3,198 @@
...
-        [Language("ga")] Irish,
-        [Language("ig")] Igbo,
-        [Language("ik")] Inupiaq,
```
Comment (linvi) [inline Tweetinvi.Core/Public/Models/Enum/Language.cs]: Why has it changed to `Inupiat`?

File: Tweetinvi.Core/Public/Models/Enum/Language.cs [hunk 3]
```diff
@@ -3,233 +3,198 @@
...
-        [Language("nv")] Navajo,
-        [Language("nv")] Navaho,
-        [Language("nb")] NorwegianBokmal,
-        [Language("nd")] NorthNdebele,
-        [Language("ne")] Nepali,
-        [Language("ng")] Ndonga,
-        [Language("nn")] NorwegianNynorsk,
-        [Language("no")] Norwegian,
-        [Language("ii")] Nuosu,
-        [Language("in")] Indian,
```
Comment (linvi) [inline Tweetinvi.Core/Public/Models/Enum/Language.cs]: Why removed?

File: Tweetinvi.Core/Public/Models/Enum/Language.cs [hunk 4]
```diff
@@ -3,233 +3,198 @@
...
-        [Language("gd")] ScottishGaelic,
-        [Language("gd")] Gaelic,
-        [Language("sn")] Shona,
-        [Language("si")] Sinhala,
-        [Language("si")] Sinhalese,
-        [Language("sk")] Slovak,
-        [Language("sl")] Slovene,
-        [Language("so")] Somali,
-        [Language("st")] SouthernSotho,
-        [Language("az")] SouthAzerbaijani,
```
Comment (linvi) [inline Tweetinvi.Core/Public/Models/Enum/Language.cs]: Why removed?

**Comment by linvi** [comment]
I have merged it and added some of the elements that you removed. If you think this is not correct please let me know and I will change that.

File: Tweetinvi.Core/Public/Models/Enum/Language.cs (hunk 1 above)
Comment (JoshKeegan) [inline Tweetinvi.Core/Public/Models/Enum/Language.cs]: Duplicate code of Hebrew

File: Tweetinvi.Core/Public/Models/Enum/Language.cs (hunk 2 above)
Comment (JoshKeegan) [inline Tweetinvi.Core/Public/Models/Enum/Language.cs]: Wherever a language has multiple common names, I tried to select the most commonly used one (not an exact science) but generally went with whatever the title of the Wikipedia article about the language was.
In this case that's Inupiat (https://en.wikipedia.org/wiki/Inupiat_language).

My use case requires being able to display a list of unique languages to my end-user, listed with their names in English. The name of the language that should be displayed should be whatever they're most likely to recognise, hence whatever it's commonly known as. So this is an area where your current solution wouldn't meet my requirements as there are multiple entries with the language code "ik", so no obvious way to select the most common name of the language.

Perhaps a compromise would be to list all the language names (using the array as you've done), with the most common first (to meet my use case), and prevent duplicate usage of the same language code (no need if each enum entry can have multiple language names and also as it removes the ambiguity of which enum entry should be returned as the language of a tweet).

File: Tweetinvi.Core/Public/Models/Enum/Language.cs (hunk 3 above)
Comment (JoshKeegan) [inline Tweetinvi.Core/Public/Models/Enum/Language.cs]: I believe "in" is actually a deprecated code for Indonesian, and I don't believe Indian is a language (although I'm certainly no expert!).

File: Tweetinvi.Core/Public/Models/Enum/Language.cs (hunk 4 above)
Comment (JoshKeegan) [inline Tweetinvi.Core/Public/Models/Enum/Language.cs]: Duplicate of Azerbaijani

**Comment by JoshKeegan** [comment]
See individual responses.

Overall, I'm happy for the multiple languages names per enum entry addition, but I could really do with a way to pick out the most common/recognisable one per entry.

I also think it's really worth considering making language codes unique as the only advantage to duplicate enum entries per language code is to pander to lazy developers who can't be bothered to look up the appropriate enum for whatever they know a language as. Making them unique would prevent the case where you search for one enum entry and get results with another, because the two enum entries are effectively identical (as they represent the same language code). 
This is particularly important as the behaviour of which enum entry would get returned isn't formally defined, so which one gets returned could change in a future release.

Also, I'm sure there were other minor corrections I made but can't remember now, so if the next release is based on the code you've committed it'll be worth checking the differences between our code on a language-by-language basis to make sure any little things aren't missed :)

**Comment by linvi** [comment]
> Overall, I'm happy for the multiple languages names per enum entry addition, but I could really do with a way to pick out the most common/recognisable one per entry.

So what I did in the current version is that you have access to both the `Name` and `Names` property. The `Name` contains the most common one, `Names` contains all the possible names.

Same goes for codes.

Would you suggest better naming for these properties maybe. I was thinking `CommonName` and `AllPossibleNames/AllNames`.

> Also, I'm sure there were other minor corrections I made but can't remember now, so if the next release is based on the code you've committed it'll be worth checking the differences between our code on a language-by-language basis to make sure any little things aren't missed :)

Already done ;)

**Comment by JoshKeegan** [comment]
Sorry, I must be going completely mad! Think it's time for some sleep...

Happy with the current names (Name/Names), but would be equally happy if you wanted to change them.

I've added comments to your commit for any individual entries where I wasn't sure about the changes made.

**Comment by JoshKeegan** [comment]
Looking pretty good now :+1: 

My only question is about `Chewa`, that has the first name set to "Nyanja"; should either the enum entry be `Nyanja` or the first name changed to "Chewa" so the first name and enum entry are consistent?
If there's a reason for it to stay as-is then please do so, it's just the only exception.

Other than that, this can be closed :)

**Comment by linvi** [comment]
I missed this one :s Thank you for taking time to review the code. I hope this will be useful.

**Comment by JoshKeegan** [comment]
No problem, and thanks for making the changes. 
And it certainly is useful, to me at least!

**Pull request closed without merging**